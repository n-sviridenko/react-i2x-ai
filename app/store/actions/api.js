import { createAction } from 'redux-actions';

export const REQUEST_FAIL = 'app/api/REQUEST_FAIL';

export const requestFail = createAction(REQUEST_FAIL);
