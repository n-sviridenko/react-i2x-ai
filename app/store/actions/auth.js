import { createAction } from 'redux-actions';
import { Map } from 'immutable';

export const LOGIN_REQUEST = 'app/auth/LOGIN_REQUEST';
export const LOGIN_SUCCESS = 'app/auth/LOGIN_SUCCESS';
export const LOGIN_FAILURE = 'app/auth/LOGIN_FAILURE';

export const loginRequest = createAction(LOGIN_REQUEST, (email, password) => Map({ email, password }));
export const loginSuccess = createAction(LOGIN_SUCCESS);
export const loginFailure = createAction(LOGIN_FAILURE);
