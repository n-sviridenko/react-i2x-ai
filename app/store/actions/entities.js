import { createAction } from 'redux-actions';

export const MERGE = 'app/entities/MERGE';

export const merge = createAction(MERGE);
