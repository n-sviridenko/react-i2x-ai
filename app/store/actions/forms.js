import { createAction } from 'redux-actions';

export const SUBMIT = 'app/forms/SUBMIT';

export const submit = createAction(SUBMIT, () => null, (options) => options);

export function buildSubmit(requestActionCreator, successActionType, failActionType) {
  return (values, dispatch) => new Promise((resolve, reject) => {
    const submitOptions = {
      requestAction: requestActionCreator(values),
      successActionType,
      failActionType,
      resolve,
      reject,
    };

    dispatch(submit(submitOptions));
  });
}
