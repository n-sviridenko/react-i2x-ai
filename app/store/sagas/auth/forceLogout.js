import { select, put } from 'redux-saga/effects';

import { getToken } from 'store/reducers/global';
import { unsetToken } from 'store/actions/global';

export default function* forceLogout() {
  const token = yield select(getToken);

  // in case if user was authenticated
  if (token !== null) {
    yield put(unsetToken());
  }
}
