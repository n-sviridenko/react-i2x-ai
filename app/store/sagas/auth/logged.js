import { put, call } from 'redux-saga/effects';

import * as jwt from 'utils/jwt';
import * as actions from 'store/actions/global';

function* logged({ payload }) {
  const token = payload.get('token');
  const time = yield call(Date.now);

  if (yield call(jwt.isValid, token, time)) {
    yield put(actions.setToken(token));
  }
}

export default logged;
