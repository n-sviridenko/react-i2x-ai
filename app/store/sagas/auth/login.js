import { put, call } from 'redux-saga/effects';

import sendRequest from 'store/sagas/api/sendRequest';
import * as actions from 'store/actions/auth';

export default function* login({ payload }) {
  const data = {
    email: payload.get('email'),
    password: payload.get('password'),
  };

  try {
    const res = yield call(sendRequest, 'POST', 'core/login', data);

    yield put(actions.loginSuccess(res));
  } catch (err) {
    yield put(actions.loginFailure(err));
  }
}
