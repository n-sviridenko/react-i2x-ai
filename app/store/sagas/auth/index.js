import { takeLatest } from 'redux-saga/effects';

import * as actions from 'store/actions/auth';
import login from './login';
import logged from './logged';

export function* loginWatcher() {
  yield takeLatest(actions.LOGIN_REQUEST, login);
}

export function* loggedWatcher() {
  yield takeLatest(actions.LOGIN_SUCCESS, logged);
}

export default [
  loginWatcher,
  loggedWatcher,
];
