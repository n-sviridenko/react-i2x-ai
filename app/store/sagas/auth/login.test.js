/* eslint-disable redux-saga/yield-effects */
import { put } from 'redux-saga/effects';
import { fromJS } from 'immutable';

import * as actions from 'store/actions/auth';
import login from './login';

describe('login', () => {
  let loginTask;

  beforeEach(() => {
    loginTask = login(actions.loginRequest('user@site.com', 'xxxx'));

    const callDescriptor = loginTask.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch loginSuccess if request was succeeded', () => {
    const res = fromJS({ token: 'yyyy' });
    const putDescriptor = loginTask.next(res).value;
    expect(putDescriptor).toEqual(put(actions.loginSuccess(res)));
  });

  it('should dispatch loginFailure if request was failed', () => {
    const error = new Error();
    const putDescriptor = loginTask.throw(error).value;
    expect(putDescriptor).toEqual(put(actions.loginFailure(error)));
  });
});
