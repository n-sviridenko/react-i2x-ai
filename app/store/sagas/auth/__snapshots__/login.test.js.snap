exports[`login should dispatch loginFailure if request was failed 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "POST",
      "core/login",
      Object {
        "email": "user@site.com",
        "password": "xxxx",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`login should dispatch loginSuccess if request was succeeded 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "POST",
      "core/login",
      Object {
        "email": "user@site.com",
        "password": "xxxx",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
